nhrpd/nhrp_peer.c

```diff
@@ -951,6 +960,35 @@ static void nhrp_peer_forward(struct nhrp_peer *p,
 				cie->holding_time = htons(if_ad->holdtime);
 			}
 			break;
+		case NHRP_EXTENSION_NAT_ADDRESS:
+			if(packet_types[hdr->type].type == PACKET_REQUEST) {
+				debugf(NHRP_DEBUG_COMMON,"Processing NHRP_EXTENSION_NAT_ADDRESS while forwarding the request packet");
+				proto = &pp->src_proto;
+			} else if(packet_types[hdr->type].type == PACKET_REPLY) {
+				debugf(NHRP_DEBUG_COMMON,"Processing NHRP_EXTENSION_NAT_ADDRESS while forwarding the reply packet");
+				/* For reply packet use protocol specified in CIE of mandatory part for cache lookup */
+				if(sockunion_family(&cie_protocol_mandatory) != AF_UNSPEC)
+					proto = &cie_protocol_mandatory;
+			}
+
+			if(proto) {
+				debugf(NHRP_DEBUG_COMMON,"Proto is %s", sockunion2str(proto, buf, sizeof(buf)));
+				c = nhrp_cache_get(nifp->ifp, proto, 0);
+				if(c) {
+					debugf(NHRP_DEBUG_COMMON,"c->cur.remote_nbma_natoa is %s", sockunion2str(&c->cur.remote_nbma_natoa, buf, sizeof(buf)) ? buf : "NULL");
+					if (sockunion_family(&c->cur.remote_nbma_natoa) != AF_UNSPEC) {
+						cie = nhrp_cie_push(zb, NHRP_CODE_SUCCESS,&c->cur.remote_nbma_natoa, proto);
+						if (!cie)
+							goto err;
+					}
+				} else {
```

Comment:
I don't understand why you add some more information on the hub.
on https://www.cisco.com/c/en/us/td/docs/ios-xml/ios/sec_conn_dmvpn/configuration/15-mt/sec-conn-dmvpn-15-mt-book/sec-conn-dmvpn-dt-spokes-b-nat.html, it simply asks to add nhrp extension if needed, but not to fill in nhrp extension.

"
 The hub receives the resolution request. If the spoke is behind a NAT device and there is no NAT extension, then the hub adds a NAT extension before forwarding this extension to the next node (spoke or next hop server) along the path.
"

Reply:
It looks like @reubendowle missed one change in the commit. I'll discuss with him. The actual change in hub is 

"The hub receives the resolution request. If the spoke is behind a NAT device and there is no NAT extension, then the hub adds a NAT extension before forwarding this extension to the next node (spoke or next hop server) along the path. However, if spoke already includes NAT extension header then it is just copied across and no changes are made to the extension header received from spoke"

Reply:
The change @amollad mentions is actually included. It is perhaps a bit confusing for the purposes of this review, but that change was added as a follow up commit. See 24c6333c8558aaa394f0865b20ac926bebd566a3

Reply:
> before going further, I would like to get a source of the specification you used to do nat extension.
> my reference is https://www.cisco.com/c/en/us/td/docs/ios-xml/ios/sec_conn_dmvpn/configuration/15-mt/sec-conn-dmvpn-15-mt-book/sec-conn-dmvpn-dt-spokes-b-nat.html.
> 
> so there are some items that are clearly not covered by this documentation.
From our research, cisco does not publically disclose the exact details of how the NAT extension should work. This is a proprietary extension (in fact the extension number 9 is actually defined in RFC2735 for a completely different purpose https://tools.ietf.org/html/rfc2735#section-4.2, so technically the NAT extension is violation of RFC. But that said the NAT extension is widely used and is required for interoperability in real world networks.

The main source of our information is reverse engineering of the protocol through observing packets from Cisco devices in wireshark.

> 
> if you reproduce it with a cisco version, thanks to communicate cisco version used.

We have tested against a number of different Cisco devices. These are the devices in our lab, although I think Amol has some others in his setup:
* Cisco 1921 with IOS 15.7(3)M7
* Cisco 2951 with IOS 15.6(3)M2
* Cisco 2901 with IOS 15.6(3)M2
* Cisco 819 with Version 15.6(3)M0a

Reply:
I will do a test with an our nhrp nat implementation. I followed the cisco link spec and expect that it will work.

Reply:
> I dit a test with an our nhrp nat implementation. I followed the cisco link spec and expect that it will work.
> 
> * frr implementation at hub side not behind nat device : interact ok with spoke other implementation (1)
> * frr implementation at spoke side not behind nat device: interact ok with spoke other implementation (2)
> 
> for (2), I could see that resolution reply is sent back via the hub, whereas on [0] it is written that resolution reply should be sent directly. So I guess the cisco implementations you used do not send directly resolution reply ?
> 
> [0] https://www.cisco.com/c/en/us/td/docs/ios-xml/ios/sec_conn_dmvpn/configuration/15-mt/sec-conn-dmvpn-15-mt-book/sec-conn-dmvpn-dt-spokes-b-nat.html
> 
> ```
> Spoke A parses the NHRP NAT-extension and builds a tunnel using Spoke B’s post-NAT address and replies directly to Spoke B.
> ```

Well, I am aware of this and this is not easy to address. The problem is when spoke receives the Resolution-Request, then nhrpd asks Zebra to update the routing table / neighbor table accordingly. However, before Zebra gets a chance to update all this, the Resolution-Reply has already left the Spoke using old routing entries - as a result the reply goes via hub. The only way to address this is to wait for Zebra to complete all routing table updates and then only send Resolution-Reply. We thought it's not worth the effort involved ....

Reply:
> > I dit a test with an our nhrp nat implementation. I followed the cisco link spec and expect that it will work.
> > 
> > * frr implementation at hub side not behind nat device : interact ok with spoke other implementation (1)
> > * frr implementation at spoke side not behind nat device: interact ok with spoke other implementation (2)
> > 
> > for (2), I could see that resolution reply is sent back via the hub, whereas on [0] it is written that resolution reply should be sent directly. So I guess the cisco implementations you used do not send directly resolution reply ?
> > [0] https://www.cisco.com/c/en/us/td/docs/ios-xml/ios/sec_conn_dmvpn/configuration/15-mt/sec-conn-dmvpn-15-mt-book/sec-conn-dmvpn-dt-spokes-b-nat.html
> > ```
> > Spoke A parses the NHRP NAT-extension and builds a tunnel using Spoke B’s post-NAT address and replies directly to Spoke B.
> > ```
> 
> Well, I am aware of this and this is not easy to address. The problem is when spoke receives the Resolution-Request, then nhrpd asks Zebra to update the routing table / neighbor table accordingly. However, before Zebra gets a chance to update all this, the Resolution-Reply has already left the Spoke using old routing entries - as a result the reply goes via hub. The only way to address this is to wait for Zebra to complete all routing table updates and then only send Resolution-Reply. We thought it's not worth the effort involved ....

ok, you mean the NBMA address of remote spoke may not be reachable directly. But don't you have spokes with default route, so that all remote spokes should be reachable directly ? I use following snipped of code to send to spoke instead of hub:
```
-void nhrp_peer_send(struct nhrp_peer *p, struct zbuf *zb)
+static void nhrp_peer_send_to(struct nhrp_peer *p, struct zbuf *zb, union sockunion *target)
 {
 	char buf[2][256];
 	struct nhrp_vrf *nhrp_vrf;
+	union sockunion *to;
 
 	nhrp_packet_debug(zb, "Send");
 
@@ -336,33 +337,42 @@ void nhrp_peer_send(struct nhrp_peer *p, struct zbuf *zb)
 		zlog_err("%s(): nhrp_vrf context not found", __func__);
 		return;
 	}
+	if (target)
+		to = target;
+	else
+		to = &p->vc->remote.nbma;
 
 	debugf(NHRP_DEBUG_KERNEL, "PACKET: Send %s -> %s",
 	       sockunion2str(&p->vc->local.nbma, buf[0], sizeof buf[0]),
-	       sockunion2str(&p->vc->remote.nbma, buf[1], sizeof buf[1]));
+	       sockunion2str(to, buf[1], sizeof buf[1]));
 
 	os_sendmsg(zb->head, zbuf_used(zb), p->ifp->ifindex,
-		   sockunion_get_addr(&p->vc->remote.nbma),
-		   sockunion_get_addrlen(&p->vc->remote.nbma),
+		   sockunion_get_addr(to), sockunion_get_addrlen(to),
 		   nhrp_vrf->nhrp_socket_fd);
 	zbuf_reset(zb);
 }
```
 and that patch
```
@@ -507,7 +536,14 @@ static void nhrp_handle_resolution_req(struct nhrp_packet_parser *pp)
 	}
 
 	nhrp_packet_complete(zb, hdr);
-	nhrp_peer_send(peer, zb);
+
+	/* if post-NAT set, reply directly to post-NAT ip address */
+	if (sockunion_family(&cie_nat) != AF_UNSPEC)
+		nbma_addr = &cie_nat;
+	else
+		nbma_addr = NULL;
+	nhrp_peer_send_to(peer, zb, nbma_addr);
```

Reply:
Is it ok to defer this change to a new pull-request? I will need to test it and currently not in position to test it

Reply:
This idea looks good. Although this patch does not fit with code in this pull- there is no cie_nat variable. I think the reply would need to go direct to the already set nbma_addr variable which has chosen the most likely true nbma, from one of nat extension,  ipsec parameters or the claimed nbma.

I don't have the time to check this out though. So as @amollad said, can we split this off. We are confident the code as is works in practice, as the hub always forwards the replies. Maybe a new issue could be created for this defect?